### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by Mnac99**
Server initial part
Client initial part

### CODE REVIEW

File: client/client.cpp [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+#include <QtWidgets>
+#include <QtNetwork>
+#include <QDebug>
+
+Client::Client(QWidget *parent)
+    : QDialog(parent)
+    ,hostCombo (new QComboBox)
+    ,portLineEdit(new QLineEdit)
+    ,getButton(new QPushButton(tr("Get ")))
+    , tcpSocket(new QTcpSocket(this))
```
Comment (KarineTadevosyan) [inline client/client.cpp:17]: try to keep the coding style, add spaces after comma

File: client/client.cpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        hostCombo->addItem(QString("localhost"));
+    }
+    QList<QHostAddress> ipAddressesList = QNetworkInterface::allAddresses();
+
+    for(int i = 0; i < ipAddressesList.size(); ++i)
+    {
+        if (!ipAddressesList.at(i).isLoopback())
+            hostCombo->addItem(ipAddressesList.at(i).toString());
+    }
+    for (int i = 0; i < ipAddressesList.size(); ++i)
```
Comment (KarineTadevosyan) [inline client/client.cpp]: looks like code duplicate here

File: client/client.cpp [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+void Client::read()
+{
+    in.startTransaction();
+
+
+    in >> list;
+    qDebug() << list;
+
+
+    QString l = list[0];
```
Comment (KarineTadevosyan) [inline client/client.cpp]: I think it'll be good to check if list is empty before accessing it by []operator

File: client/client.cpp [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    QUrl url = QUrl::fromUserInput( l);
+    QFile* destinationFile= new QFile(QStringLiteral("imageFile.jpg"));
+    Q_ASSUME(destinationFile->open(QFile::WriteOnly));
```
Comment (KarineTadevosyan) [inline client/client.cpp]: It's better to use if_else to handle both cases, also can be unexpected behavior if open fails. And don't forget to close file as well

File: client/client.cpp [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    in.startTransaction();
+
+
+    in >> list;
+    qDebug() << list;
+
+
+    QString l = list[0];
+    QUrl url = QUrl::fromUserInput( l);
+    QFile* destinationFile= new QFile(QStringLiteral("imageFile.jpg"));
```
Comment (KarineTadevosyan) [inline client/client.cpp]: No need to create file here, you should create and do whatever needed after receiving the response, also don't forget to delete when using pointers.

File: client/client.cpp [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    Q_ASSUME(destinationFile->open(QFile::WriteOnly));
+    QNetworkAccessManager* manager = new QNetworkAccessManager(this);
+    QNetworkReply* reply = manager->get(QNetworkRequest(url));
+    connect(reply,&QNetworkReply::readyRead,destinationFile,[destinationFile,reply](){destinationFile->write(reply->readAll());});
```
Comment (KarineTadevosyan) [inline client/client.cpp]: Here remove the lambda connection and write slot function for that, with checking if received reply is valid and safely write it in file there

File: server/server.cpp [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+#include <QTextStream>
+#include <QStringList>
+#include <QObject>
+#include <QDebug>
+
+
+
+Server::Server(QWidget *parent)
+    : QDialog(parent)
+    ,statusLabel(new QLabel)
```
Comment (KarineTadevosyan) [inline server/server.cpp:16]: same as in client constructor

File: server/server.cpp [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    clientConnection->disconnectFromHost();
+}
+
+QStringList Server::filereader()
+{
+
+    QFile myFile ("C:\\Users\\user\\Desktop\\Urlfile.txt");
+
+
+    if(!myFile.open(QIODevice::OpenModeFlag::ReadOnly))
```
Comment (KarineTadevosyan) [inline server/server.cpp]: close file later

File: server/server.cpp [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+
+Server::Server(QWidget *parent)
+    : QDialog(parent)
+    ,statusLabel(new QLabel)
+{
+    setWindowFlags(windowFlags() & ~Qt::WindowContextHelpButtonHint );
+    statusLabel-> setTextInteractionFlags(Qt::TextBrowserInteraction);
+    initServer();
+    list << filereader();
```
Comment (KarineTadevosyan) [inline server/server.cpp]: No need to read it here, read right before sending it.

File: server/server.h [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+private slots:
+    void send();
+    QStringList filereader();
+private:
+    void initServer();
+
+
+    QLabel *statusLabel = nullptr;
+    QTcpServer * tcpServer = nullptr;
+    QList<QString> list;
```
Comment (KarineTadevosyan) [inline server/server.h]: the same structure as QStringList, you can use it as in below line

File: server/server.h [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    QStringList l;
```
Comment (KarineTadevosyan) [inline server/server.h]: I don't think that this should be a member

**Review by KarineTadevosyan — CHANGES_REQUESTED** [review]

File: client/Client.pro [hunk 12]
```diff
@@ -0,0 +1,24 @@
+QT       += core gui
+QT       += network widgets
+
+greaterThan(QT_MAJOR_VERSION, 4): QT += widgets
```
Comment (sedrakGR) [inline client/Client.pro]: not needed

File: client/Client.pro [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+    main.cpp \
+    client.cpp
+
+HEADERS += \
+    client.h
+
+FORMS +=
+
+# Default rules for deployment.
+qnx: target.path = /tmp/$${TARGET}/bin
```
Comment (sedrakGR) [inline client/Client.pro]: what is this rule for on the path /tmp/...?

File: client/client.h [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+#include <QTcpSocket>
+QT_BEGIN_NAMESPACE
+class QComboBox;
+class QLabel;
+class QLineEdit;
+class QPushButton;
+class QTcpSocket;
+QT_END_NAMESPACE
+
+class Client : public QDialog
```
Comment (sedrakGR) [inline client/client.h:15]: Why is this a dialog?

File: client/client.h [hunk 15]
```diff
@@ -0,0 +1,46 @@
+#ifndef CLIENT_H
+#define CLIENT_H
+
+
+#include <QDataStream>
+#include <QDialog>
+#include <QTcpSocket>
+QT_BEGIN_NAMESPACE
```
Comment (sedrakGR) [inline client/client.h]: why do you need this qt_begin_namespace?

File: client/client.h [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+private slots:
+    void requestNew();
+    void read();
+
+    void displayError(QAbstractSocket::SocketError socketError);
+    void enableGetButton();
+
+private:
+
+    QComboBox *hostCombo = nullptr;
```
Comment (sedrakGR) [inline client/client.h]: you don't need these nullptrs when you also initialize them in constructor

File: server/server.h [hunk 17]
```diff
@@ -0,0 +1,33 @@
+#ifndef SERVER_H
+#define SERVER_H
+
+#include <QDialog>
+#include <QString>
+#include <QList>
+QT_BEGIN_NAMESPACE
```
Comment (sedrakGR) [inline server/server.h:7]: same: not sure you need this

File: client/client.cpp [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+
+Client::Client(QWidget *parent)
+    : QDialog(parent)
+    ,hostCombo (new QComboBox)
+    ,portLineEdit(new QLineEdit)
+    ,getButton(new QPushButton(tr("Get ")))
+    , tcpSocket(new QTcpSocket(this))
+{
+    setWindowFlags(windowFlags() & ~Qt::WindowContextHelpButtonHint);
+    hostCombo->setEditable(true);
```
Comment (sedrakGR) [inline client/client.cpp:22]: what is this combo for? I remember we specified the server ip manually

Comment (sedrakGR) [reply client/client.cpp:22]: Overall listing local hosts and ips in the client is useless, we don't want to use local ip as server (in future)

File: client/client.cpp [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    portLineEdit->setValidator(new QIntValidator(1, 65535, this));
+
+    auto hostLabel = new QLabel(tr("&Server name:"));
+    hostLabel->setBuddy(hostCombo);
+    auto portLabel = new QLabel(tr("&Server port:"));
+    portLabel->setBuddy(portLineEdit);
+
+    statusLabel = new QLabel(tr("This examples requires that you run the "
```
Comment (sedrakGR) [inline client/client.cpp]: Copied portion of a code, can be either removed or adjusted 'This examples requires that you run the'

File: client/client.cpp [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+                                " Server example as well."));
+    getButton->setDefault(false);
+    getButton->setEnabled(true);
+    auto quitButton = new QPushButton(tr("Quit"));
+
+    auto buttonBox = new QDialogButtonBox;
+    buttonBox->addButton(getButton, QDialogButtonBox::ActionRole);
+    buttonBox->addButton(quitButton, QDialogButtonBox::RejectRole);
+    in.setDevice(tcpSocket);
+    in.setVersion(QDataStream::Qt_4_0);
```
Comment (sedrakGR) [inline client/client.cpp]: what's the difference of qt_4 and others? if no any reasonable diff, then this is not needed

File: client/client.cpp [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+    connect(hostCombo, &QComboBox::editTextChanged,
+            this, &Client::enableGetButton);
+    connect(portLineEdit, &QLineEdit::textChanged,
+            this, &Client::enableGetButton);
+    connect(getButton, &QAbstractButton::clicked,
```
Comment (sedrakGR) [inline client/client.cpp:68]: As discussed we don't want the client to request the data one by one, so we need either a button to just connect to a server, not to request a new line, maybe naming fixes could cover this

File: client/client.cpp [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+
+    getButton->setEnabled(true);
+}
+void Client::read()
+{
+    in.startTransaction();
+
+
+    in >> list;
+    qDebug() << list;
```
Comment (sedrakGR) [inline client/client.cpp]: bad log, can be removed

File: client/client.cpp [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+
+    QFile* destinationFile= new QFile(QStringLiteral("imageFile.jpg"));
+    Q_ASSUME(destinationFile->open(QFile::WriteOnly));
+   for(int i = 0; i < list.size();++i)
```
Comment (sedrakGR) [inline client/client.cpp]: follow coding style please, space is missing before "++i"

File: client/client.cpp [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+   {
+    QString l = list[i];
+    QUrl url = QUrl::fromUserInput( l);
```
Comment (sedrakGR) [inline client/client.cpp]: coding style

File: client/client.cpp [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+
+
+    in >> list;
+    qDebug() << list;
+
+    QFile* destinationFile= new QFile(QStringLiteral("imageFile.jpg"));
+    Q_ASSUME(destinationFile->open(QFile::WriteOnly));
+   for(int i = 0; i < list.size();++i)
+   {
+    QString l = list[i];
```
Comment (sedrakGR) [inline client/client.cpp]: bad name of var

File: client/client.cpp [hunk 26]
```diff
@@ -0,0 +1,168 @@
...
+    QUrl url = QUrl::fromUserInput( l);
+
+
+
+    QNetworkAccessManager* manager = new QNetworkAccessManager(this);
```
Comment (sedrakGR) [inline client/client.cpp]: you create new network managers per each read function call, which doesn't make sense, better to init it in constructor or on tcp socket setup.

File: client/client.cpp [hunk 27]
```diff
@@ -0,0 +1,168 @@
...
+    QNetworkReply* reply = manager->get(QNetworkRequest(url));
+    connect(reply,&QNetworkReply::readyRead,destinationFile,[destinationFile,reply](){destinationFile->write(reply->readAll());});
+    connect(reply,&QNetworkReply::finished,destinationFile,&QFile::deleteLater);
```
Comment (sedrakGR) [inline client/client.cpp]: this is a problematic case, could be that you are not done with writing into image file, but you already delete the file handler.

File: client/client.h [hunk 28]
```diff
@@ -0,0 +1,46 @@
...
+    QLineEdit *portLineEdit = nullptr;
+    QLabel *statusLabel = nullptr;
+    QPushButton *getButton = nullptr;
+
+
+    QTcpSocket *tcpSocket = nullptr;
+    QDataStream in;
+    QString currentFortune;
+    QStringList list;
```
Comment (sedrakGR) [inline client/client.h]: ban names of members

File: client/client.h [hunk 29]
```diff
@@ -0,0 +1,46 @@
...
+
+    QComboBox *hostCombo = nullptr;
+    QLineEdit *portLineEdit = nullptr;
+    QLabel *statusLabel = nullptr;
+    QPushButton *getButton = nullptr;
+
+
+    QTcpSocket *tcpSocket = nullptr;
+    QDataStream in;
+    QString currentFortune;
```
Comment (sedrakGR) [inline client/client.h]: unused var

File: client/client.h [hunk 30]
```diff
@@ -0,0 +1,46 @@
+#ifndef CLIENT_H
+#define CLIENT_H
+
+
+#include <QDataStream>
+#include <QDialog>
+#include <QTcpSocket>
```
Comment (sedrakGR) [inline client/client.h:7]: you already did a forward declaration, no need to include it here

File: server/server.cpp [hunk 31]
```diff
@@ -0,0 +1,136 @@
...
+
+void Server::send()
+{
+
+    QByteArray block ;
+    QDataStream out(&block, QIODevice::WriteOnly);
+    out.setVersion(QDataStream::Qt_5_10);
+
+    out << list;
+    qDebug() << block << block.size();
```
Comment (sedrakGR) [inline server/server.cpp]: please remove this log

File: server/server.cpp [hunk 32]
```diff
@@ -0,0 +1,136 @@
...
+
+
+}
+
+void Server::send()
+{
+
+    QByteArray block ;
+    QDataStream out(&block, QIODevice::WriteOnly);
+    out.setVersion(QDataStream::Qt_5_10);
```
Comment (sedrakGR) [inline server/server.cpp]: you write with qt 5 stream, but read with qt 4 stream, there's an issue, I think you can skip these version setting calls

File: server/server.cpp [hunk 33]
```diff
@@ -0,0 +1,136 @@
...
+            clientConnection, &QObject::deleteLater);
+
+    clientConnection->write(block);
+    clientConnection->disconnectFromHost();
+}
+
+QStringList Server::filereader()
+{
+
+    QFile myFile ("C:\\Users\\user\\Desktop\\Urlfile.txt");
```
Comment (sedrakGR) [inline server/server.cpp]: preferrable to have 1. relative path, 2. path as input from a settings file or so. (for now only 1st point can be addressed)
Also please change the name from myFile to a better name, e.g. urlsFile, or something else which is more relevant

File: server/server.cpp (hunk 9 above)
Comment (sedrakGR) [reply server/server.cpp]: Skip this for now

File: server/server.cpp [hunk 34]
```diff
@@ -0,0 +1,136 @@
...
+
+
+    if(!myFile.open(QIODevice::OpenModeFlag::ReadOnly))
+    {
+        qCritical() << "can not open file";
+        qCritical() << myFile.errorString();
+
+
+    }
+    QTextStream stream(& myFile);
```
Comment (sedrakGR) [inline server/server.cpp]: bad case: file opening is failed, but you still put the stream on it... You need to stop further processing if the file opening is not successful.

File: server/server.cpp [hunk 35]
```diff
@@ -0,0 +1,136 @@
...
+
+
+    while(!stream.atEnd())
+    {
+
+        QString line = stream.readLine();
+
+
+        l.push_back(line);
```
Comment (sedrakGR) [inline server/server.cpp]: 1. why you keep this as a local member?
2. again the naming is bad

File: server/server.cpp [hunk 36]
```diff
@@ -0,0 +1,136 @@
...
+
+    }
+
```
Comment (sedrakGR) [inline server/server.cpp]: why so many new lines?

File: server/server.h [hunk 37]
```diff
@@ -0,0 +1,33 @@
...
+class QLabel;
+class QTcpServer;
+QT_END_NAMESPACE
+
+class Server : public QDialog
```
Comment (sedrakGR) [inline server/server.h:12]: why is server a dialog??

File: server/server.cpp [hunk 38]
```diff
@@ -0,0 +1,136 @@
...
+
+
+Server::Server(QWidget *parent)
+    : QDialog(parent)
+    ,statusLabel(new QLabel)
+{
+    setWindowFlags(windowFlags() & ~Qt::WindowContextHelpButtonHint );
+    statusLabel-> setTextInteractionFlags(Qt::TextBrowserInteraction);
+    initServer();
```
Comment (sedrakGR) [inline server/server.cpp:20]: assume we init the server, but file cannot be read or it has no data to share, does this server has anything to do?

File: server/server.cpp (hunk 9 above)
Comment (sedrakGR) [inline server/server.cpp]: Use assignment operator instead of stream

File: client/client.cpp [hunk 39]
```diff
@@ -0,0 +1,192 @@
...
+
+
+    for(int i = 0; i < list.size(); ++i)
+    {
+
+        QString l = list[i];
+        QUrl url = QUrl::fromUserInput( l);
+        QNetworkReply *reply = manager->get(QNetworkRequest(url));
+
+        in.startTransaction();
```
Comment (sedrakGR) [inline client/client.cpp:150]: why do you need this transaction?

File: client/client.cpp [hunk 40]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+    QString name = QString("imageFiles_%1.png").arg(QDateTime::currentMSecsSinceEpoch());
+
+    QFile destinationFile(name);
+
+    if(destinationFile.open(QFile::WriteOnly))
+    {
+
+        auto data = reply->readAll();
```
Comment (sedrakGR) [inline client/client.cpp:173]: check for errors in reply first, maybe reply did not properly return the file body.

File: client/client.h [hunk 41]
```diff
@@ -0,0 +1,50 @@
...
```
Comment (sedrakGR) [inline client/client.h:7]: remove this include from here

File: client/client.h [hunk 42]
```diff
@@ -0,0 +1,50 @@
...
+     QNetworkAccessManager* manager;
+
+    QComboBox *hostCombo ;
+    QLineEdit *portLineEdit;
+    QLabel *statusLabel ;
+    QPushButton *getButton ;
+
+
+    QTcpSocket *tcpSocket = nullptr;
+    QDataStream in;
```
Comment (sedrakGR) [inline client/client.h:44]: bad naming

**Review by sedrakGR — APPROVED** [review]
Still some improvements needed

**Pull request merged**